### ISSUE TITLE
Bug: 6.18.0 breaks when parsing `foo!--`

### ISSUE THREAD

**Issue by boris-petrov** [typescript-eslint/typescript-eslint#8201]
### Before You File a Bug Report Please Confirm You Have Done The Following...

- [X] I have tried restarting my IDE and the issue persists.
- [X] I have updated to the latest version of the packages.
- [X] I have [searched for related issues](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3Abug) and found none that matched my issue.
- [X] I have [read the FAQ](https://typescript-eslint.io/linting/troubleshooting) and my problem is not listed.

### Relevant Package

parser

### Playground Link

https://typescript-eslint.io/play/#ts=5.3.3&fileType=.tsx&code=DYUwLgBAHgXBB2BXAtgIxAJwgHwo%2BAJiAGYCW8IBEAvBAAwDcAUFAIQDU7DQA&eslintrc=N4KABGBEBOCuA2BTAzpAXGYBfEWg&tsconfig=N4KABGBEDGD2C2AHAlgGwKYCcDyiAuysAdgM6QBcYoEEkJemy0eAcgK6qoDCAFutAGsylBm3TgwAXxCSgA&tokens=false

### Repro Code

```TypeScript
let x: number | undefined = 0;
x!++;
```

### Expected Result

No parse error.

### Actual Result

```
Parsing error: Invalid left-hand side expression in unary operation 111:1 - 111:2
```

### Versions

| package                                | version |
| -------------------------------------- | ------- |
| `@typescript-eslint/eslint-plugin`     | `6.18.0` |
| `@typescript-eslint/parser`            | `6.18.0` |
| `@typescript-eslint/rule-tester`       | `6.18.0` |
| `@typescript-eslint/scope-manager`     | `6.18.0` |
| `@typescript-eslint/typescript-estree` | `6.18.0` |
| `@typescript-eslint/type-utils`        | `6.18.0` |
| `@typescript-eslint/utils`             | `6.18.0` |
| `TypeScript`                           | `5.3.3` |
| `ESLint`                               | `8.56.0` |
| `node`                                 | `21.5.0` |
